packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,335 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { TypedEventEmitter, assert } from "@fluidframework/common-utils";
+import { IConnectionDetails, IDeltaHandlerStrategy, IDeltaQueue, IDeltaQueueEvents } from "@fluidframework/container-definitions";
+import { IDocumentService } from "@fluidframework/driver-definitions";
+import { IDocumentMessage, ISequencedDocumentMessage, MessageType } from "@fluidframework/protocol-definitions";
+import { IConnectionManager, IConnectionManagerFactoryArgs } from "./contracts";
+import { DeltaManager } from "./deltaManager";
+
+const GroupedBatchOpType = "groupedBatch";
+
+/**
+ * Underneath, stores exactly the messages that are going to/from service.
+ * Key point is that anyone who wants to observe the events or see the content of the DeltaQueue
+ * will see the version that is virtualized (not what is sent over the wire).
+ *
+ * For example:
+ * - Say we have ops 1, 2, 3 and they get grouped together in a new "groupedBatch" op
+ * - Over the wire, we send just one op and the DeltaQueue will store just this single op
+ * - When other places subscribe to "push"/"op" event or call "toArray"/"peek", they will see a version that
+ * is inline with when the ops 1, 2, 3 were separate ops and will not see the "groupedBatch" op
+ */
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
+    private readonly _trackedListeners: Map<string, (...args: any[]) => void> = new Map();
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T) => T[],
+        _storeVirtualizedSequenceNumber?: (value: T) => void,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
```

Comment:
Can we deprecate these events (and access to queues)?
In a context of dealing with back compat later (i.e. assume that runtime and loader are coming from your PR only)?
I'd keep ability to pause queues (as it's hard to get rid of it) as a stand-alone API on IDM interface, but can we otherwise get rid of the queue exposure on public API? Is the only user here ScheduleManager? If so, I can probably help by prototyping (on top of your PR) if we can remove it (or its need to poke at these events)